sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs

```diff
@@ -207,15 +209,19 @@ private set
 
         internal bool AutoRenewLock => MaxAutoLockRenewalDuration > TimeSpan.Zero;
 
-        private readonly string _sessionId;
+        private readonly string[] _sessionIds;
+
+        private ConcurrentDictionary<string, ServiceBusReceiver> sessionIdsReceiverMap = new ConcurrentDictionary<string, ServiceBusReceiver>();
+
+        private ConcurrentDictionary<string, SemaphoreSlim> receiverCreationLockMap = new ConcurrentDictionary<string, SemaphoreSlim>();
```

Comment:
The processor seems to own multiple semaphore slim as in

- `MessageHandlerSemaphore`
- `MaxConcurrentAcceptSessionsSemaphore`
- `ProcessingStartStopSemaphore`
- The ones in `receiverCreationLockMap`

which are all resources that implement `IDisposable` should `ServiceBusProcessor` implement `IDisposable`?

Technically I think the `SemaphoreSlim` only needs to be disposed if the `WaitHandle` was acquired which is probably quite unlikely in this highly async code base

Reply:
Yeah we may want to consider implementing IAsyncDisposable and calling StopProcessingAsync in the DisposeAsync method.

Reply:
currently, we require users to explicitly call StopProcessingAsync to clean up resources, but I think we can look into whether the processor should handle this automatically when going out of scope.

Reply:
/cc @jsquire to get his thoughts as we have a similar pattern with the EventHubs processor.

Reply:
Krzysztof was against implementing `IDisposable` or `IAsyncDisposable` for the processor with the rationale that it should look like its intended use of a long-lived type that can be stopped and started multiple times.

The Event Hubs processor uses only the single semaphore, but does so in both the running and non-running states.   We were originally creating in Start and disposing in Stop, but we're now just willfully letting it live until the finalizer kicks in.   It's not ideal, but we've not yet been able to come up with a pattern that wasn't terrible or that wasn't based on disposable.

In this case, you seem to have quite a bit of semaphore action going on.   If you're using them only when the processor is running, can you kill them when the associated receiver is closed or when the processor stops?

Reply:
> In this case, you seem to have quite a bit of semaphore action going on. If you're using them only when the processor is running, can you kill them when the associated receiver is closed or when the processor stops?

Yep, I think that is what we are doing.